plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java

```diff
@@ -87,14 +93,91 @@ public Point getReferencePosition(IFigure parent, Viewport viewport, IGraphicalE
             result = parentVisibleArea.getCenter();
             FigureUtilities.translateToRelativeByIgnoringScrollbar(parent, result);
         }
+        GraphicalHelper.logical2screen(result, this.containerEditPart);
         EditPartQuery editPartQuery = new EditPartQuery(this.containerEditPart);
-        return editPartQuery.getSnapLocation(new CreateRequest(), result);
+        Point snapLocation = editPartQuery.getSnapLocation(new CreateRequest(), result);
+        GraphicalHelper.screen2logical(snapLocation, this.containerEditPart);
+        return snapLocation;
+    }
+
+    /**
+     * Overridden to snap after shiftng the point.
+     */
+    public Point validatePosition(IFigure parent, Rectangle bounds) {
+        Rectangle theBounds = bounds.getCopy();
+        IFigure clobber = findFigureIn(parent, bounds);
+        if (clobber != null) {
+            theBounds.setLocation(updateClobberedPosition(clobber, null));
+
+            // Snap
+            Point location = theBounds.getLocation();
+            GraphicalHelper.logical2screen(location, this.containerEditPart);
+            EditPartQuery editPartQuery = new EditPartQuery(this.containerEditPart);
+            Point snapLocation = editPartQuery.getSnapLocation(new CreateRequest(), location);
+            GraphicalHelper.screen2logical(snapLocation, this.containerEditPart);
+
+            theBounds.setLocation(snapLocation);
+
+            return validatePosition(parent, theBounds);
+        }
+        return theBounds.getLocation();
     }
 
     @Override
     public Point updateClobberedPosition(IFigure clobbered, IFigure newlyAddedChild) {
-        Point clobberedPosition = super.updateClobberedPosition(clobbered, newlyAddedChild);
+        Point clobberedPosition = computeTranslatedPoint(determineReferencePoint(clobbered), clobbered, true);
```

Comment:
document to explain that this method should be called only for the first created element.
(Indeed, we pass true to determineReferencePoint mthod that will imply that the shifting will be done with the grid width)

Reply:
done